src/elements/CommerceProduct.php

```diff
@@ -480,31 +481,63 @@ private function _parseVariants($event): void
 
     private function _inventoryUpdate($event): void
     {
+        /** @var Product $product */
+        $product = $event->element;
+
+        // Index variants by SKU for lookup:
+        $variantsBySku = ArrayHelper::index($event->contentData['variants'], 'sku');
 
         /** @var Commerce $commercePlugin */
         $commercePlugin = Commerce::getInstance();
-        $variants = $event->element->getVariants();
+        $variants = $product->getVariants();
 
+        // Queue up a changeset:
         $updateInventoryLevels = UpdateInventoryLevelCollection::make();
+
         foreach ($variants as $variant) {
-            if ($inventoryItem = $commercePlugin->getInventory()->getInventoryItemByPurchasable($variant)) {
-                /** @var InventoryLevel $firstInventoryLevel */
-                $firstInventoryLevel = $commercePlugin->getInventory()->getInventoryLevelsForPurchasable($variant)->first();
-                if ($firstInventoryLevel && $firstInventoryLevel->getInventoryLocation()) {
-                    $feedData = $event->feedData;
-                    $data = Json::decodeIfJson($event->feedData, true);
-                    $stock = $data['stock'] ?? 0;
-                    $updateInventoryLevels->push(new UpdateInventoryLevel([
-                            'type' => \craft\commerce\enums\InventoryTransactionType::AVAILABLE->value,
-                            'updateAction' => \craft\commerce\enums\InventoryUpdateQuantityType::SET,
-                            'inventoryItem' => $inventoryItem,
-                            'inventoryLocation' => $firstInventoryLevel->getInventoryLocation(),
-                            'quantity' => $stock,
-                            'note' => '',
-                        ])
-                    );
-                }
+            // Do we have a node for this variant at all?
+            if (!isset($variantsBySku[$variant->sku])) {
+                continue;
+            }
+
+            $stock = $variantsBySku[$variant->sku]['stock'] ?? null;
+
+            // What if the `stock` key wasn't in the incoming data?
+            if (is_null($stock)) {
+                Plugin::error(sprintf('No stock value was present in the import data for %s.', $variant->sku));
+
+                continue;
+            }
+
+            // Load InventoryItem model:
+            $inventoryItem = $commercePlugin->getInventory()->getInventoryItemByPurchasable($variant);
+
+            if (!$inventoryItem) {
+                // Not tracking, never mind!
+                continue;
+            }
```

Comment:
One thing I didn't consider when writing this (but I think is working as intended) is what would happen if the import itself turned on `inventoryTracked` for a variant. Given that at this stage of the import, the variant has been saved with its native attributes, the inventory item would exist already… right?

Reply:
(Yes, it does! Pushed another commit that rearranges some tests and clarifies how we're determining whether stock can/should be updated.)